### DISCUSSION TITLE
Relations lose "where()" restrictions if passed to child components

### DISCUSSION THREAD

**Discussion by flolanger** [livewire/livewire#4512] [1. Help]
Hi, I've got this query in my main Livewire component:

```
        $this->sprints = $this->project
            ->sprints()
            ->when($this->selectedYear, function ($query, $selectedYear) {
                return $query->where('year', $selectedYear);
            })
            ->with(['tickets' => function ($query) {
                $query->where('status', '!=', 0);
                $query->where('status', '!=', 3);
                }, 'tickets.editor', 'tickets.reviewer', 'tickets.sprint', 'tickets.category'])
            ->get();
```

I pass the `$sprints` variable to a Livewire child component to do stuff with it there. On the initial load of the page everything works as expected, but as soon as I refresh **anything** in the child component the eloquent collection `$sprints` magically increases in size of items because all these `where() `expressions for tickets e.g. `$query->where('status', '!=', 0);` I defined in the main component for `$sprints` just don't apply anymore.

How is this possible at this point? Livewire doesn't even rerun the code in the main component. I debugged it so I can confirm this. So `$sprints` should **always** stay the same as it isn't touched anymore at any time.

It seems that this is only the case for relation `where()` clauses as e.g. `$query->where('year', $selectedYear); ` is still applied.

Does anyone know a fix or workaround for this? It really restricts the use of nested components.

Thanks in advance! 

PHP: 7.4
Laravel: 8.79
Livewire: 2.8.2

**Comment by joshhanley** [comment]
@flolanger Livewire makes use of Laravel's queue serialiser, which doesn't track extra information like that, all it keeps track of is the relation name, so `tickets` in your example, so it will probably be loading all tickets without the filters.

I think the best option would be to create a computed property and add the extra relation conditionals in there, to ensure they are loaded properly each time.

Hope this helps!

**Reply by flolanger** [reply]
Oh thanks so much, I didn’t know that!

Can you give me an example for a computed property in this case? Do you mean a livewire computed property like in the docs of livewire or an laravel accessor?

**Reply by joshhanley** [reply]
@flolanger yeah a Livewire computed property, so like this

```php
public function getSprintsProperty() {
    return $this->project
            ->sprints()
            ->when($this->selectedYear, function ($query, $selectedYear) {
                return $query->where('year', $selectedYear);
            })
            ->with(['tickets' => function ($query) {
                $query->where('status', '!=', 0);
                $query->where('status', '!=', 3);
                }, 'tickets.editor', 'tickets.reviewer', 'tickets.sprint', 'tickets.category'])
            ->get();
}
```

You would have to remove your `$sprints` public property though otherwise they will conflict, but then in your view you can access `$this->sprints` and it will get the sprints from the computed property.

Compute properties are also memoized, so the query will only be run once per request.

**Reply by flolanger** [reply]
Hi again Josh,

I tested it now, and it doesn't seem to work sadly. I replaced the code like you said and removed the public `$sprints` property. Then I passed `$this->sprints` as parameter to my child livewire component:
`<livewire:projects.summary.project-summary-overall :sprints="$this->sprints" />`

But if I refresh the component, the same thing happens. The `where` clauses don't apply anymore. A simple button with `wire:click="$refresh"` is enough to corrupt the data. :/

**Reply by joshhanley** [reply]
@flolanger you will need to also follow the same pattern in the child component otherwise the same issue will happen. So I would probably just pass your project model to the child component and get the sprints from there.

**Reply by flolanger** [reply]
Thank you! But doesn’t this trigger another query for the same thing?

**Reply by flolanger** [reply]
It’s a lot of data, that’s why I’m trying to pass it instead of querying in each child.

**Reply by joshhanley** [reply]
@flolanger yeah it would. Do you need the full dataset in both components? If not maybe you can just run that query in the child component?

**Reply by flolanger** [reply]
@joshhanley I think I need to describe a bit more, what I'm trying to achieve:
I'm trying to build sort of a dashboard with lots of charts and average numbers that display basically always the same data (`$sprints` and `$tickets` of `$sprints`) but in different ways with different calculations. Currently, each of those charts / dashboard components are child livewire components, so I can do some dynamic actions. 

So my idea here is to have a parent / main livewire component that pulls all the needed data with some conditions (like which year is selected ...) and then give all those child components the data to display whatever they are supposed to display (calculations happen in childs).

Now, to avoid this problem, I sure could just put everything in one single livewire component. But I really really don't like to do this because it's a lot of code and would be really a mess to handle. Also I wouldn't be able to use some of these components / charts in another view (which I did already).

My question is now: How can I handle this situation? Do you have a different approach in mind here? I would love to hear it, because I'm really stuck here at the moment.

Thank you in advance for your help!

**Comment by joshhanley** [comment]
@flolanger hmm right yeah, so there is ways to do it, but you will need to add custom dehydrate and hydrate methods instead of the computed properties.

So at the moment when `$sprints` get's dehydrated, Livewire knows the `tickets` relation has been loaded, so it keeps track of `relations = ['tickets']` but that is it. Hence why when it gets re-hydrated on the next request, it just loads the tickets relation without any of the conditionals.

To get around this, you can unset the relation in a dehydrate method, and manually call the relation with the conditionals in a hydrate method yourself (instead of Livewire doing this for you).

To do this, in any of your components that will have sprints, make sure you add back in your sprints public property and remove the computed properties.

Then you will create two methods in each component `dehydrateSprints()` and `hydrateSprints()`.

In `dehydrateSprints()`, you should call `$this->sprints->unsetRelation('tickets');`

```php
public function dehydrateSprints()
{
    $this->sprints->unsetRelation('tickets');
}
```

In `hydrateSprints` load your tickets with conditionals (not sure if the below code will work exactly I just copied from above so you may need to experiment)

```php
public function hydrateSprints()
{
    $this->sprints->load([
        'tickets' => function ($query) {
                $query->where('status', '!=', 0);
                $query->where('status', '!=', 3);
                }
        ]);
}
```

Hope this helps!
